hail/python/test/hailtop/aiotools/copy_test_specs.py

```diff
@@ -534,427 +528,439 @@
   'dest_trailing_slash': False,
   'dest_type': 'noexist',
   'result': {'files': {'/keep': '', '/x': 'src/a'}},
   'src_trailing_slash': False,
   'src_type': 'file',
-  'treat_dest_as': 'target_file'},
+  'treat_dest_as': 'dest_is_target'},
  {'dest_basename': 'x',
   'dest_trailing_slash': True,
   'dest_type': 'noexist',
   'result': {'exception': 'FileNotFoundError'},
   'src_trailing_slash': True,
   'src_type': 'file',
-  'treat_dest_as': 'infer_target'},
+  'treat_dest_as': 'infer_dest'},
  {'dest_basename': 'x',
   'dest_trailing_slash': False,
   'dest_type': 'noexist',
   'result': {'exception': 'FileNotFoundError'},
   'src_trailing_slash': True,
   'src_type': 'file',
-  'treat_dest_as': 'infer_target'},
+  'treat_dest_as': 'infer_dest'},
  {'dest_basename': 'x',
   'dest_trailing_slash': True,
   'dest_type': 'noexist',
   'result': {'files': {'/keep': '', '/x/a': 'src/a'}},
   'src_trailing_slash': False,
   'src_type': 'file',
-  'treat_dest_as': 'infer_target'},
+  'treat_dest_as': 'infer_dest'},
  {'dest_basename': 'x',
   'dest_trailing_slash': False,
   'dest_type': 'noexist',
   'result': {'files': {'/keep': '', '/x': 'src/a'}},
   'src_trailing_slash': False,
   'src_type': 'file',
-  'treat_dest_as': 'infer_target'},
+  'treat_dest_as': 'infer_dest'},
  {'dest_basename': None,
   'dest_trailing_slash': True,
   'dest_type': 'file',
   'result': {'exception': 'NotADirectoryError'},
   'src_trailing_slash': True,
   'src_type': 'dir',
-  'treat_dest_as': 'target_dir'},
+  'treat_dest_as': 'dest_dir'},
  {'dest_basename': None,
   'dest_trailing_slash': False,
   'dest_type': 'file',
   'result': {'exception': 'NotADirectoryError'},
   'src_trailing_slash': True,
   'src_type': 'dir',
-  'treat_dest_as': 'target_dir'},
+  'treat_dest_as': 'dest_dir'},
  {'dest_basename': None,
   'dest_trailing_slash': True,
   'dest_type': 'file',
   'result': {'exception': 'NotADirectoryError'},
   'src_trailing_slash': False,
   'src_type': 'dir',
-  'treat_dest_as': 'target_dir'},
+  'treat_dest_as': 'dest_dir'},
  {'dest_basename': None,
   'dest_trailing_slash': False,
   'dest_type': 'file',
   'result': {'exception': 'NotADirectoryError'},
   'src_trailing_slash': False,
   'src_type': 'dir',
-  'treat_dest_as': 'target_dir'},
+  'treat_dest_as': 'dest_dir'},
  {'dest_basename': None,
   'dest_trailing_slash': True,
   'dest_type': 'file',
-  'result': {'exception': 'NotADirectoryError'},
+  'result': {'files': {'/a': 'dest/a',
+                       '/file1': 'src/a/file1',
+                       '/keep': '',
+                       '/subdir/file2': 'src/a/subdir/file2'}},
   'src_trailing_slash': True,
   'src_type': 'dir',
-  'treat_dest_as': 'target_file'},
+  'treat_dest_as': 'dest_is_target'},
  {'dest_basename': None,
   'dest_trailing_slash': False,
   'dest_type': 'file',
-  'result': {'exception': 'NotADirectoryError'},
+  'result': {'files': {'/a': 'dest/a',
+                       '/file1': 'src/a/file1',
```

Comment:
So these are a little hard to understand.  Let's expand it out:

src_type is a dir, so it will create `src/a/{file1, subdir/file2}`.  dest_type is `file`, so it will create `dest/a`.  The initial files are always called `a`.

dest_basename is none, but treat_dest_as = dest_is_target, so the command looks something like: `cp src/a/ dest` with the dest as the target.  So that will create `dest/{file1, subdir/file2}` coming from src (and `dest/a` and `/dest/keep`, which already existed.)

And that's exactly what we see.  Seems right to me.

Reply:
What does dest_type = 'dir' do then? I always assumed if dest_type is a file, then it's one actual file at dest

Reply:
`dest` is referring to a few things here, and that's probably confusing things.  There's the destination directory, `dest`, called `dest_base` in the code.  When the copy test is set up, I create `src/a` according to `src_type` (a file or directory), and I create `dest/a` depending on `dest_type` (a file or directory).  I create some data in the destination directory to test overwriting conditions.  Then I do the copy to destination location, called `dest` in the code, which is one of: `dest`, `dest/`, `dest/a`, `dest/a/`, `dest/x`, `dest/x/` depending on the value of `dest_basename` and `dest_trailing_slash`.

So, to specifically answer your question, `dest_type = 'dir'` means create `dest/a` a directory with some files in it (specifically, the destination directory gets `dest/a/file3` and `dest/a/subdir/file2`.

Reply:
One more observation that might be helpful.  In this example, `dest_basename = None`, so copy will copy the contents of `src/a` to `dest` directly.  The fact that `dest_type = 'dir'` is irrelevant because in this case the copy operation doesn't interact with `dest/a`.